### PR TITLE
[CALCITE-3721] Filter of distinct aggregate call is lost after applyi…

### PR DESCRIPTION
**Pull request by cshuo**
…ng `AggregateExpandDistinctAggregatesRule`

In `AggregateExpandDistinctAggregatesRule`, when the distinct aggregate call is rewritten using 
groupingSets, the filter of the distinct aggregate call itself is lost unexpected.

### CODE REVIEW

File: core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java [hunk 1]
```diff
@@ -1474,6 +1474,27 @@ private void basePushFilterPastAggWithGroupingSets(boolean unchanged)
...
+        + "FROM emp";
+    HepProgram program = new HepProgramBuilder()
+        .addRuleInstance(AggregateExpandDistinctAggregatesRule.INSTANCE)
+        .build();
+    sql(sql).with(program).check();
+  }
+
+  @Test public void testDistinctWithFilterAndGroupBy() {
+    final String sql = "SELECT deptno, SUM(comm), COUNT(DISTINCT comm),\n"
+        + "COUNT(DISTINCT sal) FILTER (WHERE sal > 1000)\n"
```
Comment (vlsi) [inline core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java]: Is this a minimal reproducible example?

Frankly speaking, it is really hard to follow the code, and it is sad it is the most trivial example in the PR.

I would prefer to have very simple example that makes it clear what the rule produces.

Comment (cshuo) [reply core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java]: Hi, vlsi, the mininal reproducible case is there exists two aggregation call, one is distinct aggregate call with filter, the other can be distinct agg call or not. I've update the JIRA description with a comprehensible case. And the test cases here can be simplified a little actually.

File: core/src/main/java/org/apache/calcite/rel/rules/AggregateExpandDistinctAggregatesRule.java [hunk 2]
```diff
@@ -402,15 +400,27 @@ private void rewriteUsingGroupingSets(RelOptRuleCall call,
...
             ImmutableBitSet.of(aggCall.getArgList())
-                .setIf(aggCall.filterArg, aggCall.filterArg >= 0)
-                .union(aggregate.getGroupSet()));
+                .setIf(filterArg, filterArg >= 0)
+                .union(aggregate.getGroupSet());
+        groupSetTreeSet.add(groupSet);
       }
+      Set<Integer> filterList = distinctFilterArgMap
+          .computeIfAbsent(groupSet, g -> new HashSet<>());
+      filterList.add(filterArg);
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/rel/rules/AggregateExpandDistinctAggregatesRule.java]: Add document to explain why we need a `Set` to keep the filter arg.

File: core/src/main/java/org/apache/calcite/rel/rules/AggregateExpandDistinctAggregatesRule.java [hunk 3]
```diff
@@ -432,31 +442,43 @@ private void rewriteUsingGroupingSets(RelOptRuleCall call,
...
+      for (Map.Entry<Pair<ImmutableBitSet, Integer>, Integer> entry : filters.entrySet()) {
+        final long v = groupValue(fullGroupSet, entry.getKey().left);
+        int distinctFilterArg = remap(fullGroupSet, entry.getKey().right);
+        RexNode expr = relBuilder.equals(nodeZ, relBuilder.literal(v));
+        if (distinctFilterArg > -1) {
+          // merge the filter of the distinct aggregate call itself.
+          RexBuilder rexBuilder = aggregate.getCluster().getRexBuilder();
+          expr = relBuilder.and(expr,
+              rexBuilder.makeCall(SqlStdOperatorTable.IS_TRUE,
+                  relBuilder.field(distinctFilterArg)));
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/rel/rules/AggregateExpandDistinctAggregatesRule.java]: Use `RelBuilder.call` instead of `RexBuilder.makeCall`.

File: core/src/main/java/org/apache/calcite/rel/rules/AggregateExpandDistinctAggregatesRule.java [hunk 4]
```diff
@@ -432,31 +442,43 @@ private void rewriteUsingGroupingSets(RelOptRuleCall call,
...
+        }
         nodes.add(
-            relBuilder.alias(
-                relBuilder.equals(nodeZ, relBuilder.literal(v)),
-                "$g_" + v));
+            relBuilder.alias(expr,
+            "$g_" + v + (distinctFilterArg < 0 ? "" : "_" + distinctFilterArg)));
       }
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/rel/rules/AggregateExpandDistinctAggregatesRule.java]: I would suggest to use `$g_v_f_distinctFilterArg` for the alias, the `f` means filter.

File: core/src/main/java/org/apache/calcite/rel/rules/AggregateExpandDistinctAggregatesRule.java [hunk 5]
```diff
@@ -470,21 +492,20 @@ private void rewriteUsingGroupingSets(RelOptRuleCall call,
...
       } else {
         aggregation = aggCall.getAggregation();
         newArgList = remap(fullGroupSet, aggCall.getArgList());
-        newFilterArg =
-            filters.get(
-                ImmutableBitSet.of(aggCall.getArgList())
-                    .setIf(aggCall.filterArg, aggCall.filterArg >= 0)
-                    .union(aggregate.getGroupSet()));
+        final ImmutableBitSet groupSet = ImmutableBitSet.of(aggCall.getArgList())
+            .setIf(aggCall.filterArg, aggCall.filterArg >= 0)
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/rel/rules/AggregateExpandDistinctAggregatesRule.java]: Maybe we can reuse the group set for each agg call.

**Comment by cshuo** [comment]
See [pull/1758](https://github.com/apache/calcite/pull/1758)
